### DISCUSSION TITLE
Add warning to documentatoin that result.all() and scalars.all() don't deduplicate

### DISCUSSION THREAD

**Discussion by ShmuelTreiger** [sqlalchemy/sqlalchemy#6711] [Ideas]
It would be great if you could please add a warning that result.all() and scalars.all() don't deduplicated, whereas query.all() [still?] does. At least for now, while emphasizing the migration to 2.0.

P.S. Sorry, I tried for a bit to figure out how to edit the documentation, but couldn't figure it out. Might also update [here](https://www.sqlalchemy.org/develop.html#documentation) with more specific instructions?

**Comment by CaselIT** [comment]
Hi,

Ah exception should be raised if the results need to be de-duplicated, so I think it's already covered without a warning.

**Reply by CaselIT** [reply]
Oh sorry, you meant in the documentation.

The documentation is inline with the code for the api ones, or in the docs folder in rst files.

Practically, you can usually search for a few words from the docs website and you will find where it's written

**Reply by ShmuelTreiger** [reply]
I received no such exception, hence why I'm requesting the documentation change, to save a future user from the digging I had to do.

> Practically, you can usually search for a few words from the docs website and you will find where it's written

What do you mean by "the docs website"? I tried searching github--first this repository and then this organization--for "method sqlalchemy.engine.Result.all()" and got no hits. Is there a separate docs source?

**Reply by CaselIT** [reply]
> I received no such exception

Strange, what query are you trying to run? It usually needs to de-duplicate if using a joinedload. Other than that the results should be what the db returned

> What do you mean by "the docs website"?

https://docs.sqlalchemy.org/en/14/ . 

> for "method sqlalchemy.engine.Result.all()"

That's an api, so it's in the code. You need to search for the result class definition. Or if you search for "Return all rows in a list." you get two results. Also usually github is not great at this, it's faster to clone and search using an editior or grep/rg or equivalent

The documentation generation is a classic sphinx configuration. It's organized as in countless other project that use sphinx so I'm not sure we can clarify it any better than pointing to the [sphinx quickstart](https://www.sphinx-doc.org/en/master/usage/quickstart.html)

**Reply by ShmuelTreiger** [reply]
I'll look into the documentation stuff, thanks.

The specific query is as follows (you can ignore the filters, but I included them for completeness):

```
session.execute(
    (
        select(User)
        .filter(User.is_visible)
        .join(GroupChatSubscription, GroupChatSubscription.user_id == User.id)
        .join(Message, Message.conversation_id == GroupChatSubscription.group_chat_id)
        .filter(Message.time >= GroupChatSubscription.joined)
        .filter(or_(Message.time <= GroupChatSubscription.left, GroupChatSubscription.left == None))
        .filter(Message.id > User.last_notified_message_id)
        .filter(Message.id > GroupChatSubscription.last_seen_message_id)
        .filter(Message.time < now() - timedelta(minutes=5))
        .filter(Message.message_type == MessageType.text)
    )
)
.scalars()
.unique()
```

Before it was `session.query(...).all()` which deduplicated, whereas the new query only does after swapping .all() for .unique()

**Reply by CaselIT** [reply]
My feeling is that what query did was wrong, since it did not return what the db returned.
That was one of the main reason why that behavior was removed by default, and must be explicitly opted into.

Btw there is already a documentation on this in the migration docs:
- in the table https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#migration-orm-usage
- in the section https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#joinedload-not-uniqued

it may probably be reasonable to link to that section from the result all documentation

**Reply by ShmuelTreiger** [reply]
Yup, I see it there, thanks! And yes, I think a link would be super helpful!

**Reply by ShmuelTreiger** [reply]
Made an attempt in #6712. It's incomplete because I'm not sure how to do the URL part.

Appreciate your help/responsiveness!

**Reply by CaselIT** [reply]
will look at it later. also mike commented already
